### PR TITLE
docs(local-review): add 5 new rules and simplify existing prose

### PR DESCRIPTION
**Pull request by ksk-jij**
## Summary

Updates the `local-review` skill (`/.claude/skills/local-review/SKILL.md`) with two passes:

**Pass 1 — add five review rules surfaced during PR #332 review**
- **Section G**: StrEnum for closed-set mode parameters; closed-set dispatch must use the Enum (via `match` or `ClassVar[dict]`), not a per-call string-keyed dict.
- **Section H**: deterministic-test **justification** (why is the asserted number correct?) and **completeness** (assert all deterministic outputs, not a subset).
- **Section M (new)**: Optimization Module Conventions — OMMX ingest/emit parity for `BinaryModel`-touching changes, and HUBO/QUBO restriction audit distinguishing *fundamental* vs *incidental* constraints.

**Pass 2 — compress redundant prose without losing concrete rules**
- Step 4's 13-item numbered checklist replaced with scope-by-section-header (each conditional section declares its own scope).
- Step 5's 5a/5b/5c structure merged into two bullets plus a performance-impact cue.
- Step 5.5's Iteration 1/2/3 formalism collapsed to moot / dependent / emergent plus a compact list of common root-cause patterns.
- H-bis's long prose around "what counts" / "reverse direction" tightened while keeping the backend matrix and severity thresholds.
- Section I's four ipynb bullets consolidated into one rule with three failure modes.

Every concrete rule in the pre-change version is preserved. Generic Python hygiene advice that adds no project-specific signal was dropped.

Pass 3 (follow-up commit) addressed self-review findings: added scope declaration to Section D, restored the `_`-prefix `__all__` rule and stale-imports bullet, the H-bis pure-refactor exemption, a Step 5 performance-impact bullet, and a common-root-cause-patterns line in Step 5.5.

## Changes

- 336 → 204 lines net (−132)
- 3 commits (add rules → compress → address local-review findings)

## Test plan
- [x] Ran the `local-review` skill on PR #332 using the new rules (exercised Enum guidance, deterministic-test justification, and HUBO audit sections).
- [x] Re-ran `local-review` on this branch twice (once after each pass); final run returned 0 findings.
- [ ] Sanity check: open `.claude/skills/local-review/SKILL.md` and verify Section cross-references (G → L, G → Step 5.5, Step 4 → D/H-bis/M) resolve correctly.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: .claude/skills/local-review/SKILL.md [hunk 1]
```diff
@@ -5,182 +5,145 @@ argument-hint: <target-branch (default: main)>
...
 
 ### G. Python Style
 
-- **Python 3.12+** features and syntax.
-- `X | None` instead of `Optional[X]`.
-- Extensive type annotations on all parameters, return types, and generics.
-- **All functions, methods, and classes — public and private alike — MUST have Google-style docstrings** with `Args`, `Returns`, and `Raises` sections (omitting only sections that genuinely do not apply), plus an `Example` where helpful. Type hints should be included in the docstring. Private helpers (`_foo`, `__bar`) may have compact sections but must still follow the Google-style structure. Tests are exempt from `Args`/`Returns` — a clear 1–2 line description of what the test verifies suffices (see Section H). Missing docstrings are a P2+ issue. See the `## Docstring Convention (MANDATORY)` section of `CLAUDE.md` for the authoritative convention and a full example.
-- No stale imports or dead code.
-- Consistent import style (`import qamomile.circuit as qmc`).
+- **Python 3.12+**, `X | None` (not `Optional[X]`), extensive type annotations.
```
[the remark this answers is not in the document]
Comment (ksk-jij) [reply .claude/skills/local-review/SKILL.md]: Fixed in 17196370. Updated Section G to "Python 3.11+" matching `pyproject.toml`'s `requires-python = ">=3.11"`, and noted that `enum.StrEnum` is available from 3.11 so the closed-set Enum guidance later in Section G still holds. Thanks for catching this.

File: .claude/skills/local-review/SKILL.md [hunk 2]
```diff
@@ -5,182 +5,145 @@ argument-hint: <target-branch (default: main)>
...
+qBraid is out of scope (executor-only wrapper around Qiskit, requires API key). Optional coverage may gate on `QBRAID_API_KEY`.
 
-**What counts as "algorithm or stdlib"**:
+**Required per backend** (parametrize or one test each):
 
-- New file under `qamomile/circuit/algorithm/` (e.g., a new variational ansatz, VQE component, new QAOA variant).
-- New file under `qamomile/circuit/stdlib/` (e.g., QFT, QPE, IQFT, future additions).
-- Any change to an existing algorithm/stdlib qkernel that alters its IR shape, parameter interface, or gate sequence.
+1. `transpiler.transpile(kernel, bindings=...)` returns a real `ExecutableProgram`.
+2. **Exercise BOTH paths**: `executable.sample(shots=...)` AND `executable.run(observable=...)` / `estimate()`. Sampler and estimator regress independently.
```
[the remark this answers is not in the document]
Comment (ksk-jij) [reply .claude/skills/local-review/SKILL.md]: Fixed in 17196370. Rewrote the expval-path requirement to match the actual API: `qmc.expval(q, hamiltonian)` inside the kernel makes `executable.run(executor, ...)` return an `ExpvalJob`, or for direct-circuit use the executor's `estimate(circuit, hamiltonian)` entry point. Confirmed against `qamomile/circuit/transpiler/executable.py:160-190`.

File: .claude/skills/local-review/SKILL.md [hunk 3]
```diff
@@ -5,182 +5,154 @@ argument-hint: <target-branch (default: main)>
...
+`Frontend (@qkernel) → IR → Transpiler Pipeline → Backend`
 
-Dependencies flow only downstream (left to right). Upstream (reverse) dependencies are forbidden:
+Dependencies flow only downstream. The practical rules:
 
 - **IR** depends on nothing — it is the shared data model.
 - **Frontend** depends only on IR. It MUST NOT import from Transpiler or Backend.
-- **Transpiler** depends only on IR. It MUST NOT import from Frontend.
-- **Backend** depends on IR and Transpiler public APIs. It MUST NOT import from Frontend.
+- **Transpiler passes** are IR-centric: each pass reads / writes IR and should not reach into Frontend internals. The `Transpiler` orchestrator class at the compile-entry boundary may reference Frontend types such as `QKernel` and `DecompositionConfig` (that is the expected compile-entry surface), but the individual passes under `qamomile/circuit/transpiler/passes/` should stay on IR.
```
[the remark this answers is not in the document]
Comment (ksk-jij) [reply .claude/skills/local-review/SKILL.md]: Fixed in 72d1a503. Verified against `qamomile/circuit/transpiler/passes/substitution.py:34,276` — `SubstitutionPass` imports `QKernel` under `TYPE_CHECKING` + a runtime late import and accepts it as a substitution target. Updated Section B to explicitly carve out the compile-entry / config-boundary pattern and narrowed "Frontend internals" to "Frontend implementation details (AST machinery, tracer internals, frontend-only helpers)". Good catch.

File: .claude/skills/local-review/SKILL.md [hunk 4]
```diff
@@ -5,182 +5,154 @@ argument-hint: <target-branch (default: main)>
...
+- Quantum building blocks exposed for composition by algorithm or optimization converters (QAOA / QRAO / FQAOA ansatz pieces, stdlib algorithms like QFT / QPE / IQFT, mixer / cost-Hamiltonian prep, etc.) MUST use `@qm.qkernel` (or `@qmc.qkernel`). `CompositeGate._decompose()` methods and emitter-side decomposition strategies are separate patterns (class method / strategy protocol) and are out of scope for this rule.
+- **Type annotations are mandatory** on all parameters and return types, using Qamomile types (`qm.Qubit`, `qm.Float`, `qm.UInt`, `qm.Vector[...]`, `qm.Dict[...]`) — not Python primitives.
+- Loops inside a qkernel body over quantum operations MUST use `qmc.range(...)` (project convention — the AST transformer technically also accepts plain `range(...)`, but the convention is to make quantum-side iteration explicit). Dict iteration uses `for k, v in mapping.items()`. Plain Python iteration over arbitrary iterables (`for x in some_list:`) is **not** captured as an IR `ForOperation` — restrict that form to iteration over precomputed concrete data used via closures (e.g., numpy arrays referenced from within the kernel body).
```
[the remark this answers is not in the document]
Comment (ksk-jij) [reply .claude/skills/local-review/SKILL.md]: Partially addressed in ea65a7d7. The **items** part is adopted — both `mapping.items()` and `qmc.items(mapping)` are now explicitly accepted (verified against `qamomile/circuit/algorithm/fqaoa.py:115,118`). The **range** part is kept as `qmc.range(...)`-preferred per project convention: `qmc` is the preferred `qamomile.circuit` alias, so new code should use `qmc.range(...)`. Older code using `qm.range(...)` is tolerated and not flagged by itself.

File: .claude/skills/local-review/SKILL.md [hunk 5]
```diff
@@ -5,182 +5,154 @@ argument-hint: <target-branch (default: main)>
...
 
+## Severity Levels
+
+Use these labels consistently when reporting findings (referenced by rules below and re-applied in Step 6):
+
+- **P0 — Bug**: incorrect behavior, runtime error, linear-type violation, silent data corruption, breaking change in unchanged files, mutable default arg, bare `except` swallowing, float `==` in production code, missing exception chaining.
+- **P1 — Significant Design Issue**: layer-boundary / `@qkernel` / backend-pattern violation; missing / stale / unexecuted `.ipynb`; new docs not in test patterns; incomplete modifications (changed interface, callers not updated, `__all__` stale); missing tolerance in numerical asserts; missing tests or docs for a new feature; OMMX parity break; missing cross-backend execution coverage for new algorithm/stdlib.
+- **P2 — Moderate**: missing docstrings, un-parametrized tests, missing randomization, missing edge / negative tests, missing deterministic-test justification or completeness, raw-string dispatch on closed sets, unnecessary copies in hot paths, unjustified incidental HUBO/QUBO guard, plain `range(...)` in a qkernel body (use `qmc.range(...)`).
```
[the remark this answers is not in the document]
Comment (ksk-jij) [reply .claude/skills/local-review/SKILL.md]: Partially adopted in ea65a7d7. Kept the P2 rule as `qmc.range(...)`-preferred since `qmc` is the project's preferred `qamomile.circuit` alias — new code should use `qmc.range(...)`, not `qm.range(...)`. Older code using `qm.range(...)` is tolerated and not flagged by itself, so files like `tests/circuit/visualization/test_regressions.py` won't be false-positived. The wording now states this preference explicitly.
